src/room.cpp

```diff
@@ -0,0 +1,188 @@
+
+    RakNet::RakString nick;
+    stream.Read(nick);
+
+    MacAddress preferred_mac;
+    stream.Read(preferred_mac);
+
+    // Verify if the nick is already taken.
+    const std::string nickname(nick.C_String(), nick.GetLength());
+    if (!IsValidNickname(nickname)) {
+        SendNameCollision(packet->systemAddress);
+        return;
+    }
+
+    if (preferred_mac != NoPreferredMac) {
+        // Verify if the preferred mac is available
+        if (!IsValidMacAddress(preferred_mac)) {
+            SendMacCollision(packet->systemAddress);
+            return;
+        }
+    } else {
+        // Assign a MAC address of this client automatically
+        preferred_mac = GenerateMacAddress();
+    }
+
+    // At this point the client is ready to be added to the room.
+    Member member{};
+    member.mac_address = preferred_mac;
+    member.nickname = nickname;
+    member.network_address = packet->systemAddress;
+
+    members.push_back(member);
+
+    // Notify everyone that the room information has changed.
+    BroadcastRoomInformation();
+
+    SendJoinSuccess(member);
+}
+
+void Room::ServerLoop() {
+    while (state != State::Closed) {
+        std::lock_guard<std::mutex> lock(server_mutex);
+
+        RakNet::Packet* packet = nullptr;
+        while (packet = server->Receive()) {
+            switch (packet->data[0]) {
+            case ID_ROOM_JOIN_REQUEST:
+                // Someone is trying to join the room.
+                HandleJoinRequest(packet);
+                break;
+            case ID_DISCONNECTION_NOTIFICATION:
+            case ID_CONNECTION_LOST:
+                // A client has disconnected, remove them from the members list if they had joined the room.
+                HandleClientDisconnection(packet->systemAddress);
+                break;
+            case ID_ROOM_WIFI_PACKET:
+                // Received a wifi packet, broadcast it to everyone else except the sender.
+                // TODO(Subv): Maybe change this to a loop over `members`, since we only want to
+                // send this data to the people who have actually joined the room.
+                server->Send(reinterpret_cast<char*>(packet->data), packet->length,
+                             HIGH_PRIORITY, RELIABLE, 0, packet->systemAddress, true);
+                break;
+            default:
+                break;
+            }
+
+            server->DeallocatePacket(packet);
+        }
+    }
+}
+
+bool Room::IsValidNickname(const std::string& nickname) {
+    // A nickname is valid if it is not already taken by anybody else in the room.
+    // TODO(Subv): Check for spaces in the name.
+
+    for (const Member& member : members) {
+        if (member.nickname == nickname) {
+            return false;
+        }
+    }
+
+    return true;
+}
+
+bool Room::IsValidMacAddress(const MacAddress& address) {
+    // A MAC address is valid if it is not already taken by anybody else in the room.
```

Comment:
If my suggestion about NoPreferredMac=FF:FF:FF:FF:FF:FF isn't considered here should't be a check that address not equals the broadcast address. I can't think of any other special macs but broadcast shoudn't be vaild